File: TheVault/Sources/VaultCore/Stats/StatsEngine.swift
import Foundation

/// Aggregations over a collection of `RunRecord`s.
///
/// Pure value-type, no I/O. Feed it `runs`, get back tables. Each summary keeps
/// `runs` (total observed), `wins`, and a derived `winrate` so callers can render
/// any view they want without re-walking the history.
public struct StatsReport: Codable {

    public struct Bucket: Codable, Hashable {
        public let key: String
        public let runs: Int
        public let wins: Int
        public var winrate: Double { runs == 0 ? 0 : Double(wins) / Double(runs) }
        public var pickedRate: Double?     // for relics: how often did this relic appear, given a baseline
    }

    public var totalRuns: Int
    public var totalWins: Int
    public var byCharacter: [Bucket]
    public var byAscension: [Bucket]
    public var byRelic: [Bucket]
    public var byArchetypeTag: [Bucket]    // optional, only filled when caller supplies tagger
    public var topPickedCards: [Bucket]
    public var topSkippedCards: [Bucket]
    public var generatedAt: Date

    public var overallWinrate: Double {
        totalRuns == 0 ? 0 : Double(totalWins) / Double(totalRuns)
    }
}

public enum StatsEngine {

    /// Build a stats report. `relicMinSample` and `cardMinSample` filter out long-tail
    /// noise so a relic seen in 1 run doesn't show as "100% winrate".
    public static func summarize(
        runs: [RunRecord],
        relicMinSample: Int = 3,
        cardMinSample: Int = 3,
        topN: Int = 15,
        archetypeTagger: ((RunRecord) -> [String])? = nil
    ) -> StatsReport {

        let total = runs.count
        let wins = runs.filter { $0.won == true }.count

        let byCharacter = bucket(
            runs: runs,
            key: { $0.character?.rawValue },
            won: { $0.won == true }
        )

        let byAscension = bucket(
            runs: runs,
            key: { $0.ascension.map { "A\($0)" } },
            won: { $0.won == true }
        )

        let byRelic = relicBuckets(
            runs: runs,
            minSample: relicMinSample,
            topN: topN
        )

        let byArchetype: [StatsReport.Bucket]
        if let tagger = archetypeTagger {
            byArchetype = archetypeBuckets(runs: runs, tagger: tagger, topN: topN)
        } else {
            byArchetype = []
        }

        let (picked, skipped) = cardPickStats(
            runs: runs,
            minSample: cardMinSample,
            topN: topN
        )

        return StatsReport(
            totalRuns: total,
            totalWins: wins,
            byCharacter: byCharacter,
            byAscension: byAscension,
            byRelic: byRelic,
            byArchetypeTag: byArchetype,
            topPickedCards: picked,
            topSkippedCards: skipped,
            generatedAt: Date()
        )
    }

    // MARK: - Internals

    private static func bucket(
        runs: [RunRecord],
        key: (RunRecord) -> String?,
        won: (RunRecord) -> Bool
    ) -> [StatsReport.Bucket] {
        var counts: [String: (runs: Int, wins: Int)] = [:]
        for r in runs {
            guard let k = key(r) else { continue }
            var c = counts[k] ?? (0, 0)
            c.runs += 1
            if won(r) { c.wins += 1 }
            counts[k] = c
        }
        return counts
            .map { StatsReport.Bucket(key: $0.key, runs: $0.value.runs, wins: $0.value.wins, pickedRate: nil) }
            .sorted { ($0.runs, $0.wins) > ($1.runs, $1.wins) }
    }

    private static func relicBuckets(
        runs: [RunRecord],
        minSample: Int,
        topN: Int
    ) -> [StatsReport.Bucket] {
        // For each relic we count: how many runs it appeared in, of which how many won.
        // We also keep an "appeared in" rate so the UI can sort by exposure as well as winrate.
        var seen: [String: (runs: Int, wins: Int)] = [:]
        for r in runs {
            let unique = Set(r.relics)
            for relic in unique {
                var s = seen[relic] ?? (0, 0)
                s.runs += 1
                if r.won == true { s.wins += 1 }
                seen[relic] = s
            }
        }
        let total = max(runs.count, 1)
        return seen
            .filter { $0.value.runs >= minSample }
            .map { entry in
                StatsReport.Bucket(
                    key: entry.key,
                    runs: entry.value.runs,
                    wins: entry.value.wins,
                    pickedRate: Double(entry.value.runs) / Double(total)
                )
            }
            .sorted { lhs, rhs in
                if lhs.winrate != rhs.winrate { return lhs.winrate > rhs.winrate }
                return lhs.runs > rhs.runs
            }
            .prefix(topN)
            .map { $0 }
    }

    private static func archetypeBuckets(
        runs: [RunRecord],
        tagger: (RunRecord) -> [String],
        topN: Int
    ) -> [StatsReport.Bucket] {
        var seen: [String: (runs: Int, wins: Int)] = [:]
        for r in runs {
            for tag in Set(tagger(r)) {
                var s = seen[tag] ?? (0, 0)
                s.runs += 1
                if r.won == true { s.wins += 1 }
                seen[tag] = s
            }
        }
        return seen
            .map { StatsReport.Bucket(key: $0.key, runs: $0.value.runs, wins: $0.value.wins, pickedRate: nil) }
            .sorted { $0.runs > $1.runs }
            .prefix(topN)
            .map { $0 }
    }

    private static func cardPickStats(
        runs: [RunRecord],
        minSample: Int,
        topN: Int
    ) -> (picked: [StatsReport.Bucket], skipped: [StatsReport.Bucket]) {
        // For each card we count: how many times it was OFFERED across all picks,
        // and how many times it was actually PICKED. "skip rate" = offered - picked.
        var offered: [String: Int] = [:]
        var picked: [String: Int] = [:]
        var pickedWins: [String: Int] = [:]
        for r in runs {
            let won = r.won == true
            for choice in r.cardPicks {
                for option in Set(choice.offered) {
                    offered[option, default: 0] += 1
                }
                if let pickedCard = choice.picked {
                    picked[pickedCard, default: 0] += 1
                    if won { pickedWins[pickedCard, default: 0] += 1 }
                }
            }
        }

        // Most-picked: weight by pick count, derive winrate from runs that actually picked.
        let mostPicked: [StatsReport.Bucket] = picked
            .filter { $0.value >= minSample }
            .map {
                StatsReport.Bucket(
                    key: $0.key,
                    runs: $0.value,
                    wins: pickedWins[$0.key] ?? 0,
                    pickedRate: Double($0.value) / Double(max(offered[$0.key] ?? 1, 1))
                )
            }
            .sorted { ($0.runs, $0.winrate) > ($1.runs, $1.winrate) }
            .prefix(topN)
            .map { $0 }

        // Most-skipped: cards offered enough to be statistically meaningful
        // AND skipped at least half the time. A card picked 100% of the time
        // it was offered isn't skipped — it just never refused — so we exclude
        // those from this view to keep the table focused.
        let mostSkipped: [StatsReport.Bucket] = offered
            .filter { $0.value >= minSample }
            .compactMap { entry -> StatsReport.Bucket? in
                let p = picked[entry.key] ?? 0
                let rate = Double(p) / Double(entry.value)
                guard rate < 0.5 else { return nil }
                return StatsReport.Bucket(
                    key: entry.key,
                    runs: entry.value,            // offered count
                    wins: p,                       // picked count (re-using `wins` slot)
                    pickedRate: rate
                )
            }
            .sorted { lhs, rhs in
                if (lhs.pickedRate ?? 0) != (rhs.pickedRate ?? 0) {
                    return (lhs.pickedRate ?? 0) < (rhs.pickedRate ?? 0)
                }
                return lhs.runs > rhs.runs   // tie-breaker: more often offered first
            }
            .prefix(topN)
            .map { $0 }

        return (mostPicked, mostSkipped)
    }
}

File: Web/lib/stats-engine.js
// stats-engine.js
// =========================================================================
// Direct JS port of TheVault/Sources/VaultCore/Stats/StatsEngine.swift.
// Same input shape (a list of RunRecord objects), same bucket math, same
// thresholds, same sort. Output shape mirrors `StatsReport` so the rendering
// layer doesn't have to think about which engine produced it.
//
// Pure functions — zero side effects, no I/O. Safe to run on the main thread
// for histories well into the thousands. (We did some napkin math: 10k runs
// × ~50 cards each × ~5 ascensions averages ≈ a few million ops, all object
// keys, all native — comfortably <100ms even on mid-range phones.)
// =========================================================================

const VALID_CHARACTERS = new Set(["ironclad", "silent", "regent", "necrobinder", "defect"]);

/**
 * Normalize a single record so downstream math doesn't have to care about the
 * dozen ways a save file might disagree with itself ("Ironclad" vs "ironclad",
 * `won: 1` vs `won: true`, missing keys, etc).
 */
export function normalizeRun(run) {
  if (!run || typeof run !== "object") return null;
  const character = (() => {
    const raw = run.character;
    if (typeof raw !== "string") return null;
    const k = raw.toLowerCase().replace(/\s+/g, "");
    if (k === "ironclad" || k === "ic") return "ironclad";
    if (k === "silent"   || k === "si") return "silent";
    if (k === "regent"   || k === "re" || k === "rg") return "regent";
    if (k === "necrobinder" || k === "nb" || k === "binder") return "necrobinder";
    if (k === "defect"   || k === "df" || k === "de") return "defect";
    return null;
  })();
  return {
    id: String(run.id ?? `${run.seed ?? "noseed"}-${run.endedAt ?? run.parsedAt ?? Date.now()}`),
    character,
    ascension: Number.isFinite(run.ascension) ? Math.max(0, Math.min(20, run.ascension | 0)) : null,
    seed: typeof run.seed === "string" ? run.seed : null,
    won: run.won === true || run.won === 1,
    floorReached: Number.isFinite(run.floorReached) ? run.floorReached | 0 : null,
    playTimeSeconds: Number.isFinite(run.playTimeSeconds) ? run.playTimeSeconds | 0 : null,
    startedAt: parseDate(run.startedAt),
    endedAt: parseDate(run.endedAt) ?? parseDate(run.parsedAt),
    relics: Array.isArray(run.relics) ? run.relics.map(String) : [],
    deckAtEnd: Array.isArray(run.deckAtEnd) ? run.deckAtEnd.map(String) : [],
    cardPicks: Array.isArray(run.cardPicks) ? run.cardPicks : [],
  };
}

function parseDate(d) {
  if (!d) return null;
  const t = Date.parse(d);
  return Number.isFinite(t) ? new Date(t) : null;
}

/**
 * Mirror of `StatsEngine.summarize` in Swift.
 */
export function summarize(runs, opts = {}) {
  const relicMinSample = opts.relicMinSample ?? 3;
  const cardMinSample  = opts.cardMinSample  ?? 3;
  const topN           = opts.topN           ?? 15;

  const total = runs.length;
  const wins  = runs.reduce((acc, r) => acc + (r.won === true ? 1 : 0), 0);

  const byCharacter = bucket(
    runs,
    (r) => r.character,
    (r) => r.won === true,
  );

  const byAscension = bucket(
    runs,
    (r) => r.ascension == null ? null : `A${r.ascension}`,
    (r) => r.won === true,
  ).sort((a, b) => parseAsc(a.key) - parseAsc(b.key));

  const byRelic = relicBuckets(runs, relicMinSample, topN);

  const { picked, skipped } = cardPickStats(runs, cardMinSample, topN);

  return {
    totalRuns: total,
    totalWins: wins,
    overallWinrate: total === 0 ? 0 : wins / total,
    byCharacter,
    byAscension,
    byRelic,
    topPickedCards: picked,
    topSkippedCards: skipped,
    generatedAt: new Date().toISOString(),
  };
}

function parseAsc(key) {
  const n = Number((key ?? "").replace(/^A/, ""));
  return Number.isFinite(n) ? n : -1;
}

function bucket(runs, keyFn, wonFn) {
  const counts = new Map();
  for (const r of runs) {
    const k = keyFn(r);
    if (k == null) continue;
    const c = counts.get(k) ?? { runs: 0, wins: 0 };
    c.runs += 1;
    if (wonFn(r)) c.wins += 1;
    counts.set(k, c);
  }
  const out = [];
  for (const [key, c] of counts) {
    out.push({
      key,
      runs: c.runs,
      wins: c.wins,
      winrate: c.runs === 0 ? 0 : c.wins / c.runs,
      pickedRate: null,
    });
  }
  return out.sort((a, b) => (b.runs - a.runs) || (b.wins - a.wins));
}

/**
 * Wilson lower-bound at 95% confidence. Penalizes small samples so a
 * 3-run / 2-win relic (66.7% raw) doesn't out-rank a 30-run / 17-win
 * relic (56.7% raw) just because the small sample lucked out. Used for
 * both ranking and color-tone in the relic and card panels — keeps the
 * "top winrate" lists honest when total runs are low.
 */
function wilsonLowerBound(wins, runs) {
  if (runs <= 0) return 0;
  const z = 1.96;
  const p = wins / runs;
  const denom = 1 + (z * z) / runs;
  const center = p + (z * z) / (2 * runs);
  const margin = z * Math.sqrt((p * (1 - p) + (z * z) / (4 * runs)) / runs);
  return Math.max(0, (center - margin) / denom);
}

function relicBuckets(runs, minSample, topN) {
  const seen = new Map();
  for (const r of runs) {
    const unique = new Set(r.relics ?? []);
    for (const relic of unique) {
      const c = seen.get(relic) ?? { runs: 0, wins: 0 };
      c.runs += 1;
      if (r.won === true) c.wins += 1;
      seen.set(relic, c);
    }
  }
  const total = Math.max(runs.length, 1);
  const out = [];
  for (const [key, c] of seen) {
    if (c.runs < minSample) continue;
    const winrate = c.runs === 0 ? 0 : c.wins / c.runs;
    out.push({
      key,
      runs: c.runs,
      wins: c.wins,
      winrate,
      lb: wilsonLowerBound(c.wins, c.runs),
      pickedRate: c.runs / total,
    });
  }
  return out
    .sort((a, b) => (b.lb - a.lb) || (b.winrate - a.winrate) || (b.runs - a.runs))
    .slice(0, topN);
}

function cardPickStats(runs, minSample, topN) {
  const offered = new Map();
  const picked = new Map();
  const pickedWins = new Map();

  const inc = (m, k) => m.set(k, (m.get(k) ?? 0) + 1);

  for (const r of runs) {
    const won = r.won === true;
    for (const choice of r.cardPicks ?? []) {
      const opts = Array.isArray(choice.offered) ? new Set(choice.offered) : null;
      if (opts) for (const o of opts) inc(offered, String(o));
      if (typeof choice.picked === "string" && choice.picked.length) {
        inc(picked, choice.picked);
        if (won) inc(pickedWins, choice.picked);
      }
    }
  }

  // Most picked: weighted by raw pick count, with derived winrate.
  const mostPicked = [];
  for (const [k, n] of picked) {
    if (n < minSample) continue;
    const w = pickedWins.get(k) ?? 0;
    const off = offered.get(k) ?? n;
    mostPicked.push({
      key: k,
      runs: n,
      wins: w,
      winrate: n === 0 ? 0 : w / n,
      lb: wilsonLowerBound(w, n),
      pickedRate: n / Math.max(off, 1),
    });
  }
  mostPicked.sort((a, b) => b.runs - a.runs || b.winrate - a.winrate);

  // Most skipped: offered enough to be meaningful AND picked < 50% of the time.
  const mostSkipped = [];
  for (const [k, off] of offered) {
    if (off < minSample) continue;
    const p = picked.get(k) ?? 0;
    const rate = p / off;
    if (rate >= 0.5) continue;
    mostSkipped.push({
      key: k,
      runs: off,
      wins: p,
      winrate: 0,
      pickedRate: rate,
    });
  }
  mostSkipped.sort((a, b) => (a.pickedRate ?? 0) - (b.pickedRate ?? 0) || b.runs - a.runs);

  return {
    picked: mostPicked.slice(0, topN),
    skipped: mostSkipped.slice(0, topN),
  };
}

/**
 * Filter helpers — replicate the macOS app's RunFilter.
 */
export function applyFilter(runs, filter = {}) {
  const { character, won, since } = filter;
  const sinceMs = since ? Date.parse(since) : null;
  return runs.filter((r) => {
    if (character && r.character !== character) return false;
    if (won === true  && r.won !== true)  return false;
    if (won === false && r.won === true)  return false;
    if (sinceMs && r.endedAt && r.endedAt.getTime() < sinceMs) return false;
    return true;
  });
}

/**
 * Pull a clean list of normalized runs out of whatever shape `history.json` is on disk.
 * Tolerates {header, runs}, {runs}, and bare arrays.
 */
export function extractRuns(parsed) {
  let arr;
  if (Array.isArray(parsed)) arr = parsed;
  else if (Array.isArray(parsed?.runs)) arr = parsed.runs;
  else if (Array.isArray(parsed?.history)) arr = parsed.history;
  else return { ok: false, error: "Couldn't find a run list in this file." };

  const runs = arr.map(normalizeRun).filter(Boolean);
  return { ok: true, runs };
}

export const VALID_CHAR_KEYS = VALID_CHARACTERS;
